File: elixir/bob/bob.exs
defmodule Bob do
  def hey(input) do
    input = input |> String.trim
    cond do
      input |> isEmpty -> "Fine. Be that way!"
      input |> isQuestion -> "Sure."
      input |> isShouting -> "Whoa, chill out!"
      true -> "Whatever."
    end
  end

  def isEmpty(input), do: input |> String.equivalent?("")
  def isQuestion(input), do: input |> String.ends_with?("?")
  def isUppercase(input), do: input |> String.upcase == input
  def isContainsChars(input), do: not (input |> String.match?(~r/^[\s01234567890,.-]*$/))
  def isShouting(input), do: isUppercase(input) && isContainsChars(input) 
end

File: go/bob/bob.go
package bob

import (
	"regexp"
	"strings"
)

const testVersion = 3

type RuleAndResponse struct {
	Rule     func(string) bool
	Response string
}

func Hey(input string) string {
	input = strings.Trim(input, " ")

	ruleAndResponses := [...]RuleAndResponse{
		{isEmpty, "Fine. Be that way!"},
		{isShouting, "Whoa, chill out!"},
		{isQuestion, "Sure."},
		{func(_ string) bool { return true }, "Whatever."},
	}

	for _, ruleAndResponse := range ruleAndResponses {
		if ruleAndResponse.Rule(input) {
			return ruleAndResponse.Response
		}
	}

	return ""
}

func isEmpty(input string) bool {
	return input == ""
}

func isQuestion(input string) bool {
	return strings.HasSuffix(input, "?")
}

func isShouting(input string) bool {
	isUppercase := strings.ToUpper(input) == input

	r := regexp.MustCompile("^[\\s01234567890,.-]*$")
	isContainsChars := !r.Match([]byte(input))

	return isUppercase && isContainsChars
}
